Implemented Batched Proof for multiple openings and Path Pruning

intx4 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before hitting that submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

This PR implements an optimized struct `MultiPath` which holds multiple proofs for multiple leaves. A `MultiPath` is generated via `MerkleTree::generate_multi_proof(...)` which takes an iterable of leaf indexes. `MultiPath` optimizes the proof size by compressing multiple paths using [Front Encoding](https://en.wikipedia.org/wiki/Incremental_encoding). To verify, `MultiPath` exposes the same API as `Path`, with a `verify(...)` method which takes an iterable of leaf hashes instead of a single hash. For verification, `MultiPath` optimizes runtime by using a lookup table to avoid recomputing redundant hashes of nodes.
The only caveat to `MultiPath` is that, when calling `generate_multi_proof`, the provided array of indexes is internally sorted to maximize compression efficiency. The array of leaves hashes provided at verification time should then match this order. The documentation provides an example code snippet.
Files affected:
- `src/merkle_tree/mod.rs`
- `src/merkle_tree/tests/mod.rs`

closes: #124 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

mmagician commented on src/merkle_tree/mod.rs:
```diff
@@ -181,6 +182,212 @@ impl<P: Config> Path<P> {
+
+        Ok(MultiPath {
+            leaf_indexes: indexes,
+            auth_paths_prefix_lenghts: auth_paths_prefix_lenghts,
+            auth_paths_suffixes: auth_paths_suffixes,
+            leaf_siblings_hashes: leaf_siblings_hashes,
```
You can use shorthand struct initialization here

mmagician commented:
That's a great addition, thanks! I left minor comments for now, though haven't gone over the actual algorithm yet.
Did you run any benches to compare performance to having individual paths?

Pratyush commented on src/merkle_tree/mod.rs:
```diff
@@ -7,6 +7,7 @@ use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
 use ark_std::borrow::Borrow;
 use ark_std::hash::Hash;
 use ark_std::vec::Vec;
+use std::collections::{BTreeSet, HashMap};
```
```suggestion
use ark_std::collections::{BTreeSet, HashMap};
```

Pratyush commented on src/merkle_tree/mod.rs:
````diff
@@ -106,7 +107,7 @@ pub type LeafParam<P> = <<P as Config>::LeafHash as CRHScheme>::Parameters;
 /// ```
 ///  Suppose we want to prove I, then `leaf_sibling_hash` is J, `auth_path` is `[C,D]`
-#[derive(Derivative, CanonicalSerialize, CanonicalDeserialize)]
+#[derive(PartialEq, Derivative, CanonicalSerialize, CanonicalDeserialize)]
 #[derivative(
     Clone(bound = "P: Config"),
````
```suggestion
#[derive(Derivative, CanonicalSerialize, CanonicalDeserialize)]
#[derivative(
    PartialEq(bound = "P: Config"),
    Clone(bound = "P: Config"),
```

intx4 commented:
> That's a great addition, thanks! I left minor comments for now, though haven't gone over the actual algorithm yet. Did you run any benches to compare performance to having individual paths?

I have implemented the benches to compare sequential generation of proofs vs generating a multiproof and sequential verification vs multiproof verification. There is a big increase in runtime for generating the proofs (for 2^20 leaves we go from ~400ms to generate the proofs to ~2.5s for generating the multiproof) and an increase of 46% for verification runtime (on my Thinkpad X1 gen10). I think the major bottleneck is the path compression and decompression part (I guess when verifying the optimization with the LUT amortizes the cost of having to first decompress, while in the proof generation the bottleneck of compression is more evident). The biggest saving is in the proof size though, from `O(n*logn)` to `O(polylog(n))` where n is the num of leaves.

I don't think there is a trivial way to optimize the generation of multiproof further (or at least I don't see it at the moment). We could optimize the verification by using `cfg_into_iter!` and using some crate providing a thread safe implementation of `HashMap` to distribute the verification of each path across threads (after sequential decompression).

mmagician commented on src/merkle_tree/mod.rs:
```diff
@@ -181,6 +183,212 @@ impl<P: Config> Path<P> {
+    ) -> Result<impl '_ + Iterator<Item = Vec<P::InnerDigest>>, crate::Error> {
+        // Incrementally reconstruct all the paths
+        let mut curr_path = self.auth_paths_suffixes[0].clone();
+        let mut auth_paths = (0..self.leaf_indexes.len())
+            .map(|_| Vec::with_capacity(curr_path.len()))
+            .collect::<Vec<Vec<P::InnerDigest>>>();
+        auth_paths[0] = self.auth_paths_suffixes[0].clone();
+
+        for i in 1..auth_paths.len() {
+            auth_paths[i].extend_from_slice(&curr_path[0..self.auth_paths_prefix_lenghts[i]]);
+            auth_paths[i].extend(self.auth_paths_suffixes[i].clone());
+            curr_path = auth_paths[i].clone();
+        }
+        Ok(auth_paths.into_iter())
+    }
```
These are effectively two extra loops that are not needed (i think).
Good that you have benches, because now you can test whether the following will give you any speedup:

Combining these into a single loop that avoids pre-mature allocation rather than creating empty vec of vectors and mutating each in place again in the next loop, why not start with an empty `let mut auth_paths = Vec::with_capacity(something)` and just push newly created `auth_path`?
Then you can avoid `extend_from_slice` followed directly by an `extend`

mmagician commented on src/merkle_tree/mod.rs:
```diff
@@ -181,6 +183,212 @@ impl<P: Config> Path<P> {
+            .collect::<Vec<Vec<P::InnerDigest>>>();
+        auth_paths[0] = self.auth_paths_suffixes[0].clone();
+
+        for i in 1..auth_paths.len() {
```
You can make the loop start at 0, I believe. This will come in handy later

mmagician commented on src/merkle_tree/mod.rs:
```diff
@@ -181,6 +183,212 @@ impl<P: Config> Path<P> {
+        // LookUp table to speedup computation avoid redundant hash computations
+        let mut hash_lut: HashMap<usize, P::InnerDigest> = HashMap::new();
+
+        for i in 0..self.leaf_indexes.len() {
```
This I wonder if there's any way to avoid calling the decompression step explicitly?

It looks like the loop that's called inside `decompress` is also going over all the leaf indices, just like you do in this loop inside `verify`.

intx4 commented:
@mmagician I fixed the code in `decompress` to avoid the extra loop and removed the decompress function entirely to merge the decompression step directly into verification, thus removing another loop. According to benches this has decreased the runtime of around 7% w.r.t the previous implementation with the extra loops.

mmagician reviewed (commented):
It is very strange that batch verification is slower than naive - if anything, by using the lookup table I'd expect a speedup. Were you able to identify which part of the new batch `verify` is causing this?

intx4 commented:
> It is very strange that batch verification is slower than naive - if anything, by using the lookup table I'd expect a speedup. Were you able to identify which part of the new batch `verify` is causing this?

the PR has been updated, now in the single threaded case MultiProof Verification is faster (verification test runs in ~2.8s on my thinkpad X1 Gen10 against ~3.8s of single proof verification). For reference, multiproof is atm only slower for creation (tests for generation run in ~600ms against ~400ms for single proof generation), but the proof size is smaller and verification is faster.

intx4 commented:
@mmagician the changes proposed by @Cesar199999 have been merged to the main branch of my fork

mmagician reviewed (approved):
That's great!
I think the checks are failing for the same reason as in other repos. @autquis already addressed these in this repo in https://github.com/arkworks-rs/crypto-primitives/pull/136.

intx4 commented:
> That's great! I think the checks are failing for the same reason as in other repos. @autquis already addressed these in this repo in #136.

@mmagician I followed @autquis commit to fix the tests, all good
